1_concepts/1_5_convert_cast_deref/src/main.rs

```diff
@@ -1,4 +1,120 @@
-fn main()
+use rand::prelude::*;
+use std::{ ops::Deref, borrow::Cow, cell::RefCell };
+
+
+#[ derive( Debug, PartialEq ) ]
+pub struct InvalidEmail;
+
+#[ derive( Debug, PartialEq ) ]
+pub struct EmailString< 'a >( Cow< 'a, str > );
+
+impl< 'a > EmailString< 'a >
 {
-  println!( "Implement me!" );
+  pub fn try_new< S >( value : S ) -> Result< Self, InvalidEmail >
+  where S : Into< Cow< 'a, str > > + Clone
+  {
+    if validator::validate_email( value.clone() )
+    {
+      Ok( Self( value.into() ) )
+    }
+    else 
+    {
+      Err( InvalidEmail )
+    }
+  }
 }
+
+#[ derive( Debug ) ]
+pub struct Random< T >
+{
+  pub values : [ T; 3 ],
+  _rng : RefCell< Box< ThreadRng > >
+} 
+
+impl< T > Random< T >
+{
+  pub fn new( first : T, second : T, third : T ) -> Self
+  {
+    Self 
+    {
+      values : [ first, second, third ],
+      _rng : RefCell::new( Box::new( thread_rng() ) )
+    }
+  }
+}
+
+impl< T > Deref for Random< T >
+{
+  type Target = T;
+
+  fn deref( &self ) -> &Self::Target 
+  {
+    &self.values.choose
+    (
+      &mut self._rng.borrow_mut().as_mut()
```

Comment:
I see no upper bound of range. 

Can you explain me how it is restricted?

Reply:
I used "choose" method which returns a reference to random one element